apollo-router/src/plugins/authorization/mod.rs

```diff
@@ -1,31 +1,323 @@
+        }
+    }
+
+    pub(crate) fn update_cache_key(context: &Context) {
+        let is_authenticated = context.contains_key(APOLLO_AUTHENTICATION_JWT_CLAIMS);
+
+        let request_scopes = context
+            .get_json_value(APOLLO_AUTHENTICATION_JWT_CLAIMS)
+            .and_then(|value| {
+                value.as_object().and_then(|object| {
+                    object.get("scope").and_then(|v| {
+                        v.as_str()
+                            .map(|s| s.split(' ').map(|s| s.to_string()).collect::<HashSet<_>>())
+                    })
+                })
+            });
+        let query_scopes = context.get_json_value(REQUIRED_SCOPES_KEY).and_then(|v| {
+            v.as_array().map(|v| {
+                v.iter()
+                    .filter_map(|s| s.as_str().map(|s| s.to_string()))
+                    .collect::<HashSet<_>>()
+            })
+        });
+
+        let mut scopes = match (request_scopes, query_scopes) {
+            (None, _) => vec![],
+            (_, None) => vec![],
+            (Some(req), Some(query)) => req.intersection(&query).cloned().collect(),
+        };
+        scopes.sort();
+
+        context
+            .upsert(QUERY_PLANNER_CACHE_KEY_METADATA, |mut o: Object| {
+                o.insert("is_authenticated", is_authenticated.into());
+                o.insert("scopes", scopes.into());
+                o
+            })
+            .unwrap();
+    }
+
+    pub(crate) fn filter_query(
+        key: &QueryKey,
+        schema: &Schema,
+    ) -> Result<Option<FilteredQuery>, QueryPlannerError> {
+        // we create a compiler to filter the query. The filtered query will then be used
+        // to generate selections for response formatting, to execute introspection and
+        // generating a query plan
+        let mut compiler = ApolloCompiler::new();
+        compiler.set_type_system_hir(schema.type_system.clone());
+        let _id = compiler.add_executable(&key.filtered_query, "query");
+
+        let is_authenticated = key
+            .metadata
+            .get("is_authenticated")
+            .and_then(|v| v.as_bool())
+            .unwrap_or_default();
+        let scopes = key
+            .metadata
+            .get("scopes")
+            .and_then(|v| v.as_array())
+            .map(|v| {
+                v.iter()
+                    .filter_map(|el| el.as_str().map(|s| s.to_string()))
+                    .collect::<Vec<_>>()
+            })
+            .unwrap_or_default();
+
+        let filter_res = Self::authenticated_filter_query(&compiler, is_authenticated)?;
+
+        let filter_res = match filter_res {
+            None => Self::scopes_filter_query(&compiler, &scopes).map(|opt| {
+                opt.map(|(query, paths)| (query, paths, Arc::new(Mutex::new(compiler))))
+            }),
+            Some((query, mut paths)) => {
+                if query.is_empty() {
+                    return Err(QueryPlannerError::PlanningErrors(PlanErrors {
```

Comment:
do we want to return a query planning error, or return an empty or null `data` field with an error?